[REVIEW] UCX - Using the new UCX Python bindings

madsbk opened the pull request:
This PR makes DASK use the new [UCX Python bindings](https://github.com/rapidsai/ucx-py/pull/180).

All tests in [`test_ucx.py`](https://github.com/dask/distributed/blob/master/distributed/comm/tests/test_ucx.py) works on my installation on a DGX-2 machine.

NB: the implementation depend on NumPy and Numba.

(cc @mrocklin @pentschev)

mrocklin commented on distributed/comm/ucx.py:
```diff
@@ -94,67 +99,92 @@ def peer_address(self) -> str:
+                    if is_cuda:
+                        frames.append(frame)
+                    else:
+                        frames.append(frame.tobytes())
```
I wonder if we could pass the memoryview `frame.data`, here instead.

madsbk replied:
Fixed

mrocklin commented on distributed/comm/ucx.py:
```diff
@@ -94,67 +99,92 @@ def peer_address(self) -> str:
-            ucp.destroy_ep(self._ep)
+        if self._ep is not None:
             logger.debug("Destroyed UCX endpoint")
+            IOLoop.current().spawn_callback(self._ep.signal_shutdown)
```
I see this more often today, and I think that they are equivalent as of 6.0

```suggestion
            IOLoop.current().add_callback(self._ep.signal_shutdown)
```

mrocklin commented on distributed/comm/ucx.py:
```diff
@@ -15,8 +15,13 @@
 
+# TODO: remove dependency on Numpy and Numba
+import numpy as np
+import numba.cuda
```
After going through this again, I'm curious about the use of Numpy to allocate buffers.  I was a little surprised by this, rather than using something like `memoryview`s (although presumably these are harder to create in pure Python).  Is this what you meant with this comment?

madsbk replied:
Yes, right now the new UCX-py implementation doesn't support `memoryview`, just because it is hard to make right compared to `__array_interface__`. But I will add it soon :)

mrocklin commented on distributed/comm/tests/test_ucx.py:
```diff
@@ -264,42 +258,35 @@ def test_ucx_deserialize():
     with LocalCluster(
-        2,
+        n_workers=2,
+        protocol="ucx://",
```
Note, you may not have intended this protocol change

madsbk replied:
Fixed

madsbk commented on distributed/comm/ucx.py:
```diff
@@ -94,67 +99,92 @@ def peer_address(self) -> str:
         serializers=("cuda", "dask", "pickle", "error"),
         on_error: str = "message",
     ):
-        with log_errors():
```
Right, removed it as part of a debugging. Fixed.

quasiben commented:
I tested this branch along with https://github.com/rapidsai/ucx-py/pull/180 and received errors when starting up a worker with dask-cuda:

## Setup
```
$ dask-scheduler --interface ib0 --protocol ucx
$ UCX_RNDV_SCHEME=put_zcopy UCX_MEMTYPE_CACHE=n UCX_TLS=rc,cuda_copy,cuda_ipc dask-cuda-worker ucx://10.33.225.165:8786
```

## Error
```
[1568841311.699848] [dgx15:40788:0]      rdmacm_ep.c:52   UCX  ERROR rdma_create_id() failed: Permission denied
distributed.nanny - ERROR - Failed to start worker
Traceback (most recent call last):
  File "/home/nfs/bzaitlen/GitRepos/distributed/distributed/nanny.py", line 674, in run
    await worker
  File "/home/nfs/bzaitlen/GitRepos/distributed/distributed/worker.py", line 1021, in start
    await self._register_with_scheduler()
  File "/home/nfs/bzaitlen/GitRepos/distributed/distributed/worker.py", line 795, in _register_with_scheduler
    self.scheduler.address, connection_args=self.connection_args
  File "/home/nfs/bzaitlen/GitRepos/distributed/distributed/comm/core.py", line 215, in connect
    quiet_exceptions=EnvironmentError,
  File "/home/nfs/bzaitlen/GitRepos/distributed/distributed/comm/ucx.py", line 207, in connect
    ep = await ucp.create_endpoint(ip, port)
  File "/home/nfs/bzaitlen/GitRepos/ucx-py/ucp/public_api.py", line 52, in create_endpoint
    return await _ctx.create_endpoint(ip_address, port)
  File "ucp/_libs/ucp_tiny.pyx", line 179, in create_endpoint
  File "ucp/_libs/ucp_tiny.pyx", line 20, in ucp._libs.ucp_tiny.assert_ucs_status
ucp.exceptions.UCXError: Input/output error
```

quasiben commented:
I spent just a bit more time looking and the top of the error has this:

```
/home/nfs/bzaitlen/miniconda3/envs/dask-ucx/lib/python3.7/contextlib.py:119: UserWarning: Creating scratch directories is taking a surprisingly long time. This is often due to running workers on a network file system. Consider specifying
a local-directory to point workers to write scratch data to a local disk.
  next(self.gen)
[UCX/ERROR] Call to cuInit results in CUDA_ERROR_NOT_INITIALIZED
[UCX/ERROR] Unable to start CUDA Context
Traceback (most recent call last):
  File "/home/nfs/bzaitlen/miniconda3/envs/dask-ucx/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 231, in initialize
    self.cuInit(0)
  File "/home/nfs/bzaitlen/miniconda3/envs/dask-ucx/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 294, in safe_cuda_api_call
    self._check_error(fname, retcode)
  File "/home/nfs/bzaitlen/miniconda3/envs/dask-ucx/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 329, in _check_error
    raise CudaAPIError(retcode, msg)
numba.cuda.cudadrv.driver.CudaAPIError: [3] Call to cuInit results in CUDA_ERROR_NOT_INITIALIZED

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/nfs/bzaitlen/GitRepos/dask-cuda/dask_cuda/initialize_context.py", line 25, in <module>
    numba.cuda.current_context()
  File "/home/nfs/bzaitlen/miniconda3/envs/dask-ucx/lib/python3.7/site-packages/numba/cuda/cudadrv/devices.py", line 213, in get_context
    return _runtime.get_or_create_context(devnum)
  File "/home/nfs/bzaitlen/miniconda3/envs/dask-ucx/lib/python3.7/site-packages/numba/cuda/cudadrv/devices.py", line 139, in get_or_create_context
    return self._get_or_create_context_uncached(devnum)
  File "/home/nfs/bzaitlen/miniconda3/envs/dask-ucx/lib/python3.7/site-packages/numba/cuda/cudadrv/devices.py", line 152, in _get_or_create_context_uncached
    with driver.get_active_context() as ac:
  File "/home/nfs/bzaitlen/miniconda3/envs/dask-ucx/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 387, in __enter__
    driver.cuCtxGetCurrent(byref(hctx))
  File "/home/nfs/bzaitlen/miniconda3/envs/dask-ucx/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 274, in __getattr__
    self.initialize()
  File "/home/nfs/bzaitlen/miniconda3/envs/dask-ucx/lib/python3.7/site-packages/numba/cuda/cudadrv/driver.py", line 234, in initialize
    raise CudaSupportError("Error at driver init: \n%s:" % e)
numba.cuda.cudadrv.error.CudaSupportError: Error at driver init:
[3] Call to cuInit results in CUDA_ERROR_NOT_INITIALIZED:
distributed.preloading - INFO - Import preload module: dask_cuda.initialize_context
```

dask-cuda runs the following on startup: https://github.com/rapidsai/dask-cuda/blob/a05de107bbfef30a87c98b933110ac354fef21dd/dask_cuda/initialize_context.py#L18-L24

Is this cuda context management strategy no longer necessary with the new python bindings ?

madsbk commented:
Thanks for the hint @quasiben. I think we still need the dask-cuda startup script in order to support multiple threads, but based one your observation I found and fixed the problem in UCX-Py. 
I fixed the problem by delaying the instantiation of UCX-Py to the first API use.

quasiben commented:
@madsbk That fixed the issue!  Thanks!

quasiben commented on distributed/comm/ucx.py:
```diff
@@ -95,66 +98,89 @@ def peer_address(self) -> str:
     ):
         with log_errors():
+            if self.closed():
+                raise CommClosedError
```
Maybe this could have a more descriptive message ?

> CommClosedError("Endpoint is closed -- unable to send message)

quasiben commented on distributed/comm/ucx.py:
```diff
@@ -95,66 +98,89 @@ def peer_address(self) -> str:
+                    await self._ep.shutdown()
+                    self._ep.close()
+                self._ep = None
+                raise CommClosedError()
```
Could this also have a more descriptive message:

> CommClosedError("Endpoint have closed -- shutting down)

mrocklin commented:
Checking in here.  Dask is likely to release in the next couple of days.  Should we merge this in now, or wait for the next release in a couple of weeks?

madsbk commented:
> Checking in here. Dask is likely to release in the next couple of days. Should we merge this in now, or wait for the next release in a couple of weeks?

I think there is two blockers:

- The default branch of https://github.com/rapidsai/ucx-py should be changed to `master`.
- Final test on a DGX machine

madsbk commented:
@mrocklin, I think this PR is ready for merge now.

mrocklin commented:
I plan to release this project today.  I am inclined to do the release, and then merge this afterwards.  Do you agree or should this go in this release?  We will likely release again soon.

madsbk commented:
> I plan to release this project today. I am inclined to do the release, and then merge this afterwards. Do you agree or should this go in this release? We will likely release again soon.

I am not sure but it would be great if it could be release before RAPIDS v0.10.

mrocklin commented:
Alright.  Merging in.  This should be in the 2.5 release.
